versioned/transfer/src/main/java/org/projectnessie/versioned/transfer/CommitLogOptimization.java

```diff
@@ -0,0 +1,272 @@
+                  .flatMap(Collection::stream))) {
+        entries
+            .map(
+                entry -> {
+                  try {
+                    return databaseAdapter.rebuildKeyList(entry, h -> null);
+                  } catch (ReferenceNotFoundException ignore) {
+                    // Ignore this, because it can only happen if `entry`, which has been read
+                    // before, no longer exists.
+                    return null;
+                  }
+                })
+            .filter(Objects::nonNull)
+            .forEach(
+                e -> {
+                  try {
+                    databaseAdapter.updateMultipleCommits(singletonList(e));
+                  } catch (ReferenceNotFoundException ex) {
+                    throw new RuntimeException(ex);
+                  }
+                });
+      }
+    }
+  }
+
+  /**
+   * Keeps track of the recent N ({@code parents-per-commit + 1}) commit log entries to populate the
+   * lists of parent commits, both {@link #handleEntry(CommitLogEntry) while scanning} the commit
+   * log and when {@link #drain() reaching the "beginning of time"}.
+   */
+  static final class CommitParentsState {
+
+    private final DatabaseAdapter databaseAdapter;
+    private final int parentsPerCommit;
+    private final ArrayDeque<CommitLogEntry> lastEntries;
+
+    CommitParentsState(DatabaseAdapter databaseAdapter) {
+      this.databaseAdapter = databaseAdapter;
+      this.parentsPerCommit = databaseAdapter.getConfig().getParentsPerCommit();
+      this.lastEntries = new ArrayDeque<>(this.parentsPerCommit + 1);
+    }
+
+    boolean canStopIterating(CommitLogEntry entry) {
+      return entry.getParents().size() >= parentsPerCommit;
+    }
+
+    private boolean full() {
+      return lastEntries.size() == this.parentsPerCommit + 1;
+    }
+
+    void handleEntry(CommitLogEntry entry) {
+      if (full()) {
+        lastEntries.removeFirst();
+      }
+      lastEntries.add(entry);
+      if (full()) {
+        updateLeastRecentCommitLogEntryParents();
+      }
+    }
+
+    void drain() {
+      while (!lastEntries.isEmpty()) {
+        updateLeastRecentCommitLogEntryParents();
+      }
+    }
+
+    private void updateLeastRecentCommitLogEntryParents() {
+      CommitLogEntry commitToUpdate = lastEntries.removeFirst();
+      List<Hash> currentParents = commitToUpdate.getParents();
+      if (currentParents.size() >= parentsPerCommit) {
+        return;
+      }
+
+      ImmutableCommitLogEntry.Builder newEntry =
+          ImmutableCommitLogEntry.builder().from(commitToUpdate);
+
+      List<Hash> calculatedParents = new ArrayList<>(lastEntries.size());
+      lastEntries.forEach(e -> calculatedParents.add(e.getHash()));
+      if (calculatedParents.size() < parentsPerCommit) {
+        calculatedParents.add(databaseAdapter.noAncestorHash());
+      }
+
+      if (!calculatedParents.subList(0, currentParents.size()).equals(currentParents)) {
```

Comment:
`currentParents.size()` may be larger than `calculatedParents.size()` in theory. Should it be `Math.min(calculatedParents.size(), currentParents.size())`?

Reply:
It can't be - see line 239

Reply:
Yeah, sorry about the false alarm :facepalm: